src/privileged/messagingSystem/api.js

```diff
@@ -0,0 +1,171 @@
+/* eslint no-console: ["warn", { allow: ["info", "warn", "error"] }] */
+/* global ExtensionAPI */
+
+"use strict";
+
+this.messagingSystem = class extends ExtensionAPI {
+  getAPI(context) {
+    const { Services } = ChromeUtils.import(
+      "resource://gre/modules/Services.jsm",
+      {},
+    );
+
+    const { ExtensionCommon } = ChromeUtils.import(
+      "resource://gre/modules/ExtensionCommon.jsm",
+      {},
+    );
+
+    const { EventManager } = ExtensionCommon;
+
+    const { ExtensionUtils } = ChromeUtils.import(
+      "resource://gre/modules/ExtensionUtils.jsm",
+      {},
+    );
+    const { ExtensionError } = ExtensionUtils;
+
+    const { RemoteSettings } = ChromeUtils.import(
+      "resource://services-settings/remote-settings.js",
+      {},
+    );
+    const { ASRouterTargeting } = ChromeUtils.import(
+      "resource://activity-stream/lib/ASRouterTargeting.jsm",
+      {},
+    );
+    const { MessageLoaderUtils } = ChromeUtils.import(
+      "resource://activity-stream/lib/ASRouter.jsm",
+      {},
+    );
+
+    const PREF_ASROUTER_CFR_PROVIDER =
+      "browser.newtabpage.activity-stream.asrouter.providers.cfr";
+
+    const getUpdatedCfrProviderPref = (
+      bucket,
+      cohort,
+      personalizedModelVersion = null,
+    ) => {
+      const existingValue = Services.prefs.getStringPref(
```

Comment:
This works, you could also consider using `defineLazyPreferenceGetter` to manage the parsing etc. https://searchfox.org/mozilla-central/source/js/xpconnect/loader/XPCOMUtils.jsm#341

Reply:
That's good to know. Leaving as is for now but it may be useful in the future, thanks :)